### PR TITLE
[ENH]: int / float-tuple like kwarg legend(loc) for rcParams['legend.loc']

### LINKED ISSUE #22338: [Bug]: rcParams['legend.loc'] can't use float-tuple like kwarg legend(loc...)
### Bug summary

We can now use a (x,y) argument for `plt.legend(loc=(x,y))`,
but using the tuple as a value for rcParams['legend.loc'] fails. 

### Code for reproduction

```python
import matplotlib.pyplot as plt
import matplotlib as mpl
xs, ys = [1,2,3], [2,3,1]

fig, ax = plt.subplots(3)

ax[0].scatter(xs, ys, label='loc-tuple-arg')
ax[0].legend(loc=(1.1, .5)) # works

ax[1].scatter(xs, ys, label='loc-rcparam-tuple')
mpl.rcParams['legend.loc'] = (0.9, .7)
ax[1].legend() # fails w/o documentation

ax[2].scatter(xs, ys, label='loc-rcparam-str(tuple)')
mpl.rcParams['legend.loc'] = '(.8, .3)'
ax[2].legend() # fails with documentation
```

### Actual outcome

Setting rcParams['legend.loc'] to a tuple fails with 

> AttributeError: 'tuple' object has no attribute 'lower'

setting rcParams['legend.loc'] to the string representation of a tuple fails with a list of supported values. 

### Expected outcome

Ideally  rcParams['legend.loc']  would apply a tuple the same way the kwarg does. 

Second best would be a bit more documentation; 1) reporting the list of supported strings for any un-interpretable value, and 2) mentioning in the default matplotlibrc  that only a set of strings are supported. 

### Additional information

I believe the  `loc` kwarg originally only took string values, but that seems to have been at least a major version ago. 

### Operating system

Ubuntu

### Matplotlib Version

3.3.4

### Matplotlib Backend

TkAgg

### Python version

Python 3.9.7

### Jupyter version

_No response_

### Installation

pip

### PR DESCRIPTION
**Pull request by NoyHanan**
closes #22338 

## PR summary
I Added the ability to use int / float-tuple like kwarg legend(loc) for rcParams['legend.loc'], by creating a new validator for rcParams['legend.loc'].
This new change will help to a consistent and a uniform approach to using the legend.
Addresses both parts of the issue

## PR checklist

- [x] "closes #0000" is in the body of the PR description to [link the related issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [x] new and changed code is [tested](https://matplotlib.org/devdocs/devel/testing.html)
- [N/A] *Plotting related* features are demonstrated in an [example](https://matplotlib.org/devdocs/devel/documenting_mpl.html#writing-examples-and-tutorials)
- [x] *New Features* and *API Changes* are noted with a [directive and release note](https://matplotlib.org/devdocs/devel/coding_guide.html#new-features-and-api-changes)
- [x] Documentation complies with [general](https://matplotlib.org/devdocs/devel/documenting_mpl.html#writing-rest-pages) and [docstring](https://matplotlib.org/devdocs/devel/documenting_mpl.html#writing-docstrings) guidelines

### CODE REVIEW

**Comment by chahak13** [comment]
Thank you for your first PR 🎉  Looking forward to more such contributions! :)

**Comment by NoyHanan** [comment]
> Thank you for your first PR 🎉 Looking forward to more such contributions! :)

Thank you so much!

File: lib/matplotlib/rcsetup.py
Comment (NoyHanan) [reply lib/matplotlib/rcsetup.py]: We sure can! thanks for teaching me something new, I didn't we had that option in python :) 
I updated the code.

**Comment by NoyHanan** [comment]
@chahak13 Do I need to note for the API change / add documentation?

**Comment by chahak13** [comment]
There should be a note in the what's new section I believe, yes.

File: lib/matplotlib/tests/test_rcparams.py
Comment (NoyHanan) [reply lib/matplotlib/tests/test_rcparams.py:603]: Yes, I made sure of it.

File: lib/matplotlib/rcsetup.pyi
Comment (tacaswell) [inline lib/matplotlib/rcsetup.pyi:160]: Should this be 
```suggestion
def validate_legend_loc(loc: str | int | tuple[float, float]
```

and the same on the return?

File: lib/matplotlib/tests/test_rcparams.py
Comment (tacaswell) [inline lib/matplotlib/tests/test_rcparams.py:608]: If we are going to make it take these values at run time in Python we also need to be able to put them in the matplotribrc file so if it is a string we need to take a try at the "sequence of floats" validator as well.

File: doc/api/next_api_changes/behavior/22338-NH.rst
Comment (tacaswell) [inline doc/api/next_api_changes/behavior/22338-NH.rst:4]: This should be in `whats_new` not API change (this is a new feature that we did not used to have, not a change to existing behavior that might break a user).

**Comment by tacaswell** [comment]
Thank you for your work on this @NoyHanan , I think it is headed the right direction!  However I left a couple of comments that need to be addressed.

File: lib/matplotlib/rcsetup.pyi
Comment (ksunden) [reply lib/matplotlib/rcsetup.pyi:160]: Actually, validators in this file are currently type hinted as `Any` for the input, as the whole point is to pass unknown input and determine if they are valid for the expected types.

We use (and have for a long time) exceptions here. If we had instead used boolean-returning functions, we could potentially use [PEP 647 TypeGuards](https://peps.python.org/pep-0647/) for this idea, but with exceptions, this is pretty much how to do it (as exceptions are not type hinted)

But the point about it being a 2-tuple of float is valid.

Comment (ksunden) [reply lib/matplotlib/rcsetup.pyi:160]: (With the asterisk on using PEP 647 being that it is python version 3.10+, but as bool is not how these work, that is a moot point anyway)

**Comment by NoyHanan** [comment]
> Thank you for your work on this @NoyHanan , I think it is headed the right direction! However I left a couple of comments that need to be addressed.

Hi @tacaswell,
Thank you for taking the time to review my PR. I appreciate your feedback and will begin working on making the necessary changes.

File: lib/matplotlib/tests/test_rcparams.py
Comment (NoyHanan) [reply lib/matplotlib/tests/test_rcparams.py:608]: I don't want to waste your time, but I'm not sure I fully understood what you're saying. Can you guide me a bit more?

Comment (NoyHanan) [reply lib/matplotlib/tests/test_rcparams.py:608]: @tacaswell Please don't forget about me :)

Comment (tacaswell) [reply lib/matplotlib/tests/test_rcparams.py:608]: sorry, I get too many notifications.....


I mean that anything in the rcParams should be setabble from `matplotlibrc`.  Given that the matplotlibrc is a simple text format the values come to us ...as text so the validator (which is more of a converter, but that is a different topic), needs to correctly convert a sequence of floats the way that is encoded in `matplotlibrc`.

**Comment by tacaswell** [comment]
There should also be a test of reading an the values from a file.

File: lib/matplotlib/tests/test_rcparams.py
Comment (NoyHanan) [reply lib/matplotlib/tests/test_rcparams.py:608]: All good, I had no doubt about it.
That helped a lot! thank you so much. I'll start working on it.

**Comment by NoyHanan** [comment]
> There should also be a test of reading an the values from a file.

On it.

**Comment by NoyHanan** [comment]
Hi @tacaswell!
I know that you many notifications to attend to, so I wanted to kindly inform you that I have made the necessary adjustments to the code based on your feedback (or at least I believe I have :) ).

**Comment by NoyHanan** [comment]
> Hi @tacaswell! I know that you many notifications to attend to, so I wanted to kindly inform you that I have made the necessary adjustments to the code based on your feedback (or at least I believe I have :) ).

@tacaswell please don't forget about me :) :)

File: lib/matplotlib/rcsetup.py
Comment (rcomer) [inline lib/matplotlib/rcsetup.py:765]: The value does not need to be positive here.  We can set negative numbers to have a legend outside of the axes:

```python
fig, ax = plt.subplots(layout="constrained")
ax.plot([0, 1], label="foo")
ax.legend(loc=[-0.3, 0.2])
plt.show()
```
image

File: lib/matplotlib/rcsetup.py
Comment (rcomer) [inline lib/matplotlib/rcsetup.py:726]: I think this should be `versionadded` rather than `versionchanged`, since it is a new function.  Should probably also update the version to 3.8 since this PR is in that milestone.

File: lib/matplotlib/rcsetup.py
Comment (rcomer) [inline lib/matplotlib/rcsetup.py:751]: I think the int handling should go below the string handling:  if the user sets the int in their matplotlibrc, similar to the tuple case, it would need converting from string to int before checking.

File: lib/matplotlib/tests/test_rcparams.py
Comment (rcomer) [inline lib/matplotlib/tests/test_rcparams.py:609]: Worth adding an int value here I think.

**Review by rcomer — COMMENTED** [review]
Thank you for your work on this @NoyHanan, and apologies for the delay in reviewing.  Could you update the title of this PR to something more descriptive?

**Comment by NoyHanan** [comment]
> 
Hi @rcomer, 
I appreciate your thorough review of my code. I've made the necessary corrections based on your feedback. I'm looking forward to your thoughts on the updates. Thank you!

File: lib/matplotlib/rcsetup.py
Comment (QuLogic) [inline lib/matplotlib/rcsetup.py:763]:
```suggestion
        if len(loc) == 2 and all(isinstance(e, Real) for e in loc):
```

File: lib/matplotlib/rcsetup.py
Comment (QuLogic) [inline lib/matplotlib/rcsetup.py:17]: `Real` is already imported below, this isn't needed.

File: lib/matplotlib/rcsetup.py
Comment (QuLogic) [inline lib/matplotlib/rcsetup.py:733]: I'm not sure this needs to be public.

File: doc/users/next_whats_new/rcParams[legend.loc]_supports_float_tuple.rst
Comment (QuLogic) [inline doc/users/next_whats_new/rcParams[legend.loc]_supports_float_tuple.rst:2]:
```suggestion
``rcParams['legend.loc']`` now accepts float-tuple inputs
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

File: doc/users/next_whats_new/rcParams[legend.loc]_supports_float_tuple.rst
Comment (QuLogic) [inline doc/users/next_whats_new/rcParams[legend.loc]_supports_float_tuple.rst:5]:
```suggestion
The :rc:`legend.loc` rcParams now accepts float-tuple inputs, same as the *loc* keyword argument to `.Legend`.
This allows users to set the location of the legend in a more flexible and consistent way.
```

**Review by rcomer — APPROVED** [review]
Thanks @NoyHanan, your updates look good to me.

**Comment by NoyHanan** [comment]
> Thanks @NoyHanan, your updates look good to me.

@rcomer thanks alot! a few more things:
1. I'm commiting the changes @QuLogic, does it need to be approved again?
2. who may merge the pull request to the main repo?

**Comment by rcomer** [comment]
For code changes, we require two approvals from core developers. So yes, please address @QuLogic's comments. Then once he is happy he will likely approve and merge.

File: lib/matplotlib/rcsetup.py
Comment (NoyHanan) [reply lib/matplotlib/rcsetup.py]: No problem! changed it to private.

**Comment by QuLogic** [comment]
It looks like you'll also need to update the typing stubs now that the validator is private. Also, you may want to squash the changes together.

**Comment by NoyHanan** [comment]
> It looks like you'll also need to update the typing stubs now that the validator is private. Also, you may want to squash the changes together.

Sure, no problem.
Can you explain a bit more about how to update?

**Comment by ksunden** [comment]
Just undo the change to the `.pyi` file

**Comment by NoyHanan** [comment]
> Just undo the change to the `.pyi` file

Thanks!

**Comment by rcomer** [comment]
To squash, you can do an [interactive rebase](https://matplotlib.org/devdocs/devel/development_workflow.html#rewrite-commit-history).  I strongly recommend saving a backup copy of your branch first if you have not done this (much) before.

**Comment by NoyHanan** [comment]
HI @QuLogic, I'll appriciate your review on the changes I made 🙏

**Comment by rcomer** [comment]
@NoyHanan are you happy to squash the commits, or would you prefer us to take care of that?

**Comment by NoyHanan** [comment]
> @NoyHanan are you happy to squash the commits, or would you prefer us to take care of that?

I can take care of that, I just thought you guys meant I should squash only the last 3 commits I made for @rcomer notes.
Should I sqaush them all?  and thank you so much for your help and patience.

**Comment by rcomer** [comment]
I think minimal numbers of commits are generally preferred.

**Comment by NoyHanan** [comment]
> I think minimal numbers of commits are generally preferred.

Thank you for guiding me. I hope that its fine now.

**Comment by NoyHanan** [comment]
Hi @QuLogic, could you please review the changes I made? 🙏

**Review by QuLogic — APPROVED** [review]

**Comment by rcomer** [comment]
Congratulations on your first PR merged into Matplotlib @NoyHanan!  We hope to hear from you again.

**Comment by NoyHanan** [comment]
> Congratulations on your first PR merged into Matplotlib @NoyHanan! We hope to hear from you again.

Thank you so much, I appreciate your support. You'll definitely hear from me again :)

**Pull request merged**
